prov/sm2/src/sm2_coordination.c

```diff
@@ -0,0 +1,574 @@
+					"file size is reset)!\n",
+					item);
+				strncpy(entries[item].ep_name,
+					ZOMBIE_ALLOCATION_NAME, SM2_NAME_MAX);
+				goto retry_lookup;
+			}
+		}
+
+		if (!self) {
+			if (!pid_lives(abs(entries[item].pid))) {
+				entries[item].pid = 0;
+			}
+			/* Someone else allocated the entry for us */
+			goto found;
+		}
+
+		if (entries[item].pid <= 0) {
+			if (!pid_lives(abs(entries[item].pid))) {
+				FI_WARN(&sm2_prov, FI_LOG_AV,
+					"During sm2 allocation of space for "
+					"endpoint named %s pid %d "
+					"pre-allocated space at allocation "
+					"entry[%d] and then died!\n",
+					name, -entries[item].pid, item);
+			}
+			goto found;
+		}
+
+		FI_WARN(&sm2_prov, FI_LOG_AV,
+			"During sm2 allocation of space for endpoint named %s "
+			"an existing conflicting address was found at "
+			"allocation entry[%d]\n",
+			name, item);
+
+		if (!pid_lives(entries[item].pid)) {
+			FI_WARN(&sm2_prov, FI_LOG_AV,
+				"The pid which allocated the conflicting "
+				"allocation entry is "
+				"dead. Reclaiming as our own.\n");
+			/* it is possible that EP's referencing this region are
+			 * still alive... don't know how to check (they likely
+			 * died if PID died) */
+			goto found;
+		}
+
+		FI_WARN(&sm2_prov, FI_LOG_AV,
+			"ERROR: The endpoint (pid: %d) with conflicting "
+			"address %s is still alive.\n",
+			entries[item].pid, name);
+		return -FI_EADDRINUSE;
+	}
+
+	/* fine, we could not find the entry, so now look for an empty slot */
+	for (item = 0; item < SM2_MAX_UNIVERSE_SIZE; item++) {
+		peer_pid = entries[item].pid;
+		if (peer_pid == 0)
+			goto found;
+		if (peer_pid < 0)
+			/* A third peer might have entered this address into
+			 * their AV, and there is no current way to check
+			 * this... need to keep this entry in the file until we
+			 * clean up */
+			continue;
+
+		if (!pid_lives(peer_pid)) {
+			sm2_file_extend_for_region(map, item);
+			entries = sm2_mmap_entries(map);
+			struct sm2_region *peer_region =
+				sm2_mmap_ep_region(map, item);
+
+			if (entries[item].startup_ready &&
+			    smr_freestack_isfull(sm2_free_stack(peer_region))) {
+				/* we found a slot with a dead PID and
+				 * the freestack is full */
+				entries[item].pid = 0;
+				goto found;
+			}
+		}
+	}
+
+	FI_WARN(&sm2_prov, FI_LOG_AV,
+		"No available entries were found in the coordination file, all "
+		"%d were used\n",
+		SM2_MAX_UNIVERSE_SIZE);
+	return -FI_EAVAIL;
+
+found:
+	if (self) {
+		entries[item].startup_ready = 0;
+		atomic_wmb();
+		entries[item].pid = pid;
+	}
+
+	if (!self && entries[item].pid == 0) {
+		entries[item].startup_ready = 0;
+		atomic_wmb();
+		entries[item].pid = -pid;
+	}
+
+	FI_WARN(&sm2_prov, FI_LOG_AV,
+		"Using sm2 region at allocation entry[%d] for %s\n", item,
+		name);
+
+	strncpy(entries[item].ep_name, name, SM2_NAME_MAX - 1);
+	entries[item].ep_name[SM2_NAME_MAX - 1] = '\0';
+
+	*gid = item;
+
+	sm2_file_extend_for_region(map, item);
+
+	return 0;
+}
```

Comment:
This function should be broken up into more manageable chunks.

Reply:
Agreed... but does this have to happen now?  If it does, how would you recommend that I do it?